server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java

```diff
@@ -327,13 +341,87 @@ protected void setMinSeqNoToKeep(long seqNo) {
+        super.trimUnreferencedReaders();
+
+        // cleans up remote translog files not referenced in latest uploaded metadata.
+        // This enables us to restore translog from the metadata in case of failover or relocation.
+        Set<Long> generationsToDelete = new HashSet<>();
+        for (long generation = minRemoteGenReferenced - 1; generation >= 0; generation--) {
+            if (fileTransferTracker.uploaded(Translog.getFilename(generation)) == false) {
+                break;
+            }
+            generationsToDelete.add(generation);
+        }
+        deleteRemoteGeneration(generationsToDelete);
+        deleteOlderPrimaryTranslogFilesFromRemoteStore();
+    }
+
+    private void deleteOlderPrimaryTranslogFilesFromRemoteStore() {
+        // The deletion of older translog files in remote store is on best-effort basis, there is a possibility that there
+        // are older files that are no longer needed and should be cleaned up. In here, we delete all files that are part
+        // of older primary term.
+        if (olderPrimaryCleaned.trySet(Boolean.TRUE)) {
+            logger.info("Cleaning up translog uploaded by previous primaries");
+            long minPrimaryTermInMetadata = getMinPrimaryTermInMetadata();
+            if (minPrimaryTermInMetadata == -1) {
+                // We should keep all primary terms since we can not determine the minimum primary term referenced by the metadata file
+                return;
+            }
+            Set<Long> primaryTermsInRemote = getPrimaryTermsInRemote();
+            // Delete all primary terms that are no more referenced by the metadata file and exists in the
+            Set<Long> primaryTermsToDelete = primaryTermsInRemote.stream()
+                .filter(term -> term < minPrimaryTermInMetadata)
+                .collect(Collectors.toSet());
+            primaryTermsToDelete.forEach(term -> translogTransferManager.deleteTranslogAsync(term, new ActionListener<>() {
+                @Override
+                public void onResponse(Void response) {
+                    // NO-OP
+                }
+
+                @Override
+                public void onFailure(Exception e) {
+                    logger.error(
+                        () -> new ParameterizedMessage("Exception occurred while deleting older translog files for primary_term={}", term),
+                        e
+                    );
+                }
+            }));
+        }
+    }
+
+    private Set<Long> getPrimaryTermsInRemote() {
         try {
-            translogTransferManager.deleteTranslog(primaryTermSupplier.getAsLong(), reader.generation);
-        } catch (IOException ignored) {
-            logger.error("Exception {} while deleting generation {}", ignored, reader.generation);
+            return translogTransferManager.listPrimaryTerms();
+        } catch (IOException e) {
+            logger.error("Exception occurred while getting oldest primary term in remote store", e);
```

Comment:
nit: getting **all** the primary terms from the remote store

Reply:
Incorporated